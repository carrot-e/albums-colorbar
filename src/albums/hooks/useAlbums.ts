import { useState } from 'react';
import dataProvider, { RawAlbumData } from '../data/dataProvider';
import { getPredominantColorFromImgURL } from 'color-thief-react/lib/utils';
import { HSBToRGB, RGBToHSB } from '../../utils/colors';

export default () => {
  const [loading, setLoading] = useState<boolean>(false);
  const [data, setData] = useState<Album[]>([]);

  const fetchAlbums = async () => {
    setLoading(true);

    try {
      const response = await dataProvider.getAlbums();
      const albums = await Promise.all(
        response.feed.entry.map((rawEntry, index: number) =>
          dataMapper(rawEntry, index + 1)
        )
      );
      setData(() => albums);
    } finally {
      setLoading(false);
    }
  };

  return { fetchAlbums, loading, data };
};

const normalizeDominantColor = ([r, g, b]: [number, number, number]): [
  number,
  number,
  number
] => {
  const [h] = RGBToHSB(r, g, b);
  return HSBToRGB(h, 100, 50);
};

const dataMapper = async (
  rawData: RawAlbumData,
  ranking: number
): Promise<Album> => {
  const rawImageData = rawData['im:image'][rawData['im:image'].length - 1];
  return {
    id: rawData.id.attributes['im:id'],
    name: rawData['im:name'].label,
    link: rawData.link.attributes.href,
    itemCount: rawData['im:itemCount'].label,
    image: {
      // workaround for better image quality
      src: rawImageData.label.replace('170x170bb.png', '632x632bb.webp'),
      attributes: rawImageData.attributes,
    },
    category: {
      id: rawData.category.attributes['im:id'],
      label: rawData.category.attributes.label,
    },
    releaseDate: new Date(rawData['im:releaseDate'].label),
    artist: {
      name: rawData['im:artist'].label,
    },
    rights: rawData.rights.label,
    ranking,
    dominantColor: normalizeDominantColor(
      await getPredominantColorFromImgURL(
        rawImageData.label,
        'rgbArray',
        'anonymous'
      )
    ),
  };
};

export interface Album {
  id: string;
  name: string;
  link: string;
  itemCount: string;
  image: {
    src: string;
    attributes: {
      height: string;
    };
  };
  category: {
    id: string;
    label: string;
  };
  releaseDate: Date;
  artist: {
    name: string;
  };
  ranking: number;
  dominantColor: [number, number, number];
  rights: string;
}
