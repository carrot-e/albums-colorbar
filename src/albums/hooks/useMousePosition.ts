import { useEffect, useState } from 'react';

export default () => {
  const [mousePosition, setMousePosition] = useState<{
    clientX: number;
    clientY: number;
  }>();

  const updateMousePosition = (e: MouseEvent) => {
    setMousePosition({ clientX: e.clientX, clientY: e.clientY });
  };

  useEffect(() => {
    window.addEventListener('mousemove', updateMousePosition, true);

    return () => {
      window.removeEventListener('mousemove', updateMousePosition, true);
    };
  }, []);

  return mousePosition;
};
