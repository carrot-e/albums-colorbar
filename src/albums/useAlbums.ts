import { useState } from 'react';
import dataProvider, { RawAlbumData } from './dataProvider';

export default () => {
  const [loading, setLoading] = useState<boolean>(false);
  const [data, setData] = useState<Album[]>([]);

  const fetchAlbums = async () => {
    setLoading(true);

    try {
      const response = await dataProvider.getAlbums();
      setData(() =>
        response.feed.entry.map((rawEntry, index: number) =>
          dataMapper(rawEntry, index + 1)
        )
      );
    } finally {
      setLoading(false);
    }
  };

  return { fetchAlbums, loading, data };
};

const dataMapper = (rawData: RawAlbumData, ranking: number): Album => {
  const rawImageData = rawData['im:image'][rawData['im:image'].length - 1];
  return {
    id: rawData.id.attributes['im:id'],
    name: rawData['im:name'].label,
    link: rawData.link.attributes.href,
    image: {
      src: rawImageData.label,
      attributes: rawImageData.attributes,
    },
    category: {
      id: rawData.category.attributes['im:id'],
      label: rawData.category.attributes.label,
    },
    releaseDate: new Date(rawData['im:releaseDate'].label),
    artist: {
      name: rawData['im:artist'].label,
    },
    ranking,
  };
};

export interface Album {
  id: string;
  name: string;
  link: string;
  image: {
    src: string;
    attributes: {
      height: string;
    };
  };
  category: {
    id: string;
    label: string;
  };
  releaseDate: Date;
  artist: {
    name: string;
  };
  ranking: number;
}
