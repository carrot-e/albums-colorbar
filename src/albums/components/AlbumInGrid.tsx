import {
  differenceInWeeks,
  format,
  formatDistanceToNow,
  isToday,
} from 'date-fns';
import React, { useContext } from 'react';
import { Album as AlbumInterface } from '../hooks/useAlbums';
import PageStateContext from '../context/PageStateContext';

function AlbumInGrid({ album }: { album: AlbumInterface }) {
  const { setSelectedAlbum } = useContext(PageStateContext);
  return (
    <li
      className="basis-1/2 md:basis-1/3 lg:basis-1/4 xl:basis-1/4 2xl:basis-1/5 md:p-5 p-3 cursor-pointer"
      onClick={() => setSelectedAlbum(album.id)}
    >
      <div>
        <div className="w-full">
          <img
            className="w-full rounded-md"
            src={album.image.src}
            alt={album.name}
          />
        </div>
        <div className="mt-2">
          <p className="text-md font-bold text-white">{album.name}</p>
          <p className="text-md text-white opacity-70">{album.artist.name}</p>
          {isToday(album.releaseDate) && (
            <p className="text-sm text-white opacity-50">Today</p>
          )}
          {differenceInWeeks(new Date(), album.releaseDate) >= 1 && (
            <p className="text-sm text-white opacity-50">
              {format(album.releaseDate, 'dd MMMM yyyy')}
            </p>
          )}
          {differenceInWeeks(new Date(), album.releaseDate) < 1 &&
            !isToday(album.releaseDate) && (
              <p className="text-sm text-white opacity-30">
                {formatDistanceToNow(album.releaseDate)} ago
              </p>
            )}
        </div>
      </div>
    </li>
  );
}

export default AlbumInGrid;
