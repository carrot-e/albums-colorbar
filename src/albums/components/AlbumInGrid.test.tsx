import React from 'react';
import { render, screen } from '@testing-library/react';
import { format, subDays } from 'date-fns';
import { Album as AlbumInterface } from '../hooks/useAlbums';
import AlbumInGrid from './AlbumInGrid';

const album: AlbumInterface = {
  id: '1640572729',
  name: 'Faith in the Future (Deluxe)',
  link: 'https://music.apple.com/us/album/faith-in-the-future-deluxe/1640572729?uo=2',
  image: {
    src: 'https://is4-ssl.mzstatic.com/image/thumb/Music112/v4/59/f9/f3/59f9f3b6-e815-99fd-a359-10fec504a5e3/4050538836639.jpg/170x170bb.png',
    attributes: { height: '170' },
  },
  category: {
    id: '14',
    label: 'Pop',
  },
  releaseDate: new Date('2022-10-01'),
  artist: {
    name: 'Louis Tomlinson',
  },
  ranking: 3,
  dominantColor: [0, 0, 0],
  rights: 'This Compilation â„— 2013 Capitol Records, LLC',
  itemCount: '15',
};

describe('<AlbumInGrid />', () => {
  test('renders <AlbumInGrid />', () => {
    const component = render(<AlbumInGrid album={album} />);
    expect(component).toMatchSnapshot();
  });

  test('renders <AlbumInGrid /> released today', () => {
    album.releaseDate = new Date();
    render(<AlbumInGrid album={album} />);

    expect(screen.queryByText('Today')).toBeInTheDocument();
    expect(screen.queryByText('day ago')).not.toBeInTheDocument();
  });

  test('renders <AlbumInGrid /> released during the week', () => {
    album.releaseDate = subDays(new Date(), 6);
    render(<AlbumInGrid album={album} />);

    expect(screen.queryByText('6 days ago')).toBeInTheDocument();
  });

  test('renders <AlbumInGrid /> released more than a week ago', () => {
    album.releaseDate = subDays(new Date(), 7);
    render(<AlbumInGrid album={album} />);

    expect(screen.queryByText('7 days ago')).not.toBeInTheDocument();
    expect(
      screen.queryByText(format(album.releaseDate, 'dd MMMM yyyy'))
    ).toBeInTheDocument();
  });
});
