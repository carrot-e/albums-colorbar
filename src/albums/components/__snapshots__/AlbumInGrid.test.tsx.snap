// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlbumInGrid /> renders <AlbumInGrid /> 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="basis-1/2 md:basis-1/3 lg:basis-1/4 xl:basis-1/4 2xl:basis-1/5 md:p-5 p-3 cursor-pointer"
      >
        <div>
          <div
            class="w-full"
          >
            <img
              alt="Faith in the Future (Deluxe)"
              class="w-full rounded-md"
              src="https://is4-ssl.mzstatic.com/image/thumb/Music112/v4/59/f9/f3/59f9f3b6-e815-99fd-a359-10fec504a5e3/4050538836639.jpg/170x170bb.png"
            />
          </div>
          <div
            class="mt-2"
          >
            <p
              class="text-md font-bold text-white"
            >
              Faith in the Future (Deluxe)
            </p>
            <p
              class="text-md text-white opacity-70"
            >
              Louis Tomlinson
            </p>
            <p
              class="text-sm text-white opacity-50"
            >
              01 October 2022
            </p>
          </div>
        </div>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="basis-1/2 md:basis-1/3 lg:basis-1/4 xl:basis-1/4 2xl:basis-1/5 md:p-5 p-3 cursor-pointer"
    >
      <div>
        <div
          class="w-full"
        >
          <img
            alt="Faith in the Future (Deluxe)"
            class="w-full rounded-md"
            src="https://is4-ssl.mzstatic.com/image/thumb/Music112/v4/59/f9/f3/59f9f3b6-e815-99fd-a359-10fec504a5e3/4050538836639.jpg/170x170bb.png"
          />
        </div>
        <div
          class="mt-2"
        >
          <p
            class="text-md font-bold text-white"
          >
            Faith in the Future (Deluxe)
          </p>
          <p
            class="text-md text-white opacity-70"
          >
            Louis Tomlinson
          </p>
          <p
            class="text-sm text-white opacity-50"
          >
            01 October 2022
          </p>
        </div>
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
